### ISSUE TITLE
[Feature Request][Lie] Support for SO(3) and SE(3) Lie groups

### ISSUE THREAD

**Issue by krrish94** [kornia/kornia#483]
Created this to discuss implementation of rigid-body motion parameterization on SO(3) and SE(3).

Currently, several optimization schemes for `sfm` and related problems parameterize rotations and rototranslations as Lie group elements. Implementing these utils is also part of the `kornia.sfm` roadmap (see https://github.com/kornia/kornia/issues/480).

I think it makes sense for these ops to be in `kornia.linalg` (given all the other transformation stuff is in there). For the internal design, we could have `SO(3)` utils and `SE(3)` utils inside separate submodules and implement `torch.nn.Module`s for the `ExpMap` and `LogMap` for both manifolds.

Alternate propositions welcome!

**Comment by edgarriba** [comment]
Yes, please proceed with it. In internal discussion we agreed to put geometry transforms under `kornia.geometry.linalg` so that we can clearly separate image stuff from pure maths transformations.

For the `nn.Module` what we have been trying to do all the time is for any feature to provide both module and free function. Modules are mainly designed to be a part of a `nn.Sequential` which is relatively flexible, so we might always give priority to free functions unless there's a optimization to take into account with the module state.

**Comment by krrish94** [comment]
LGTM, I'd presonally shorten the module name to `kornia.geometry.linalg.lie`

**Comment by edgarriba** [comment]
sure, go ahead. Just make sure that when you design the API, take into account namings so that functionalities could be exposed to top level.

**Comment by edgarriba** [comment]
@krrish94 any update on this ?

**Comment by versatran01** [comment]
@edgarriba I would like to help on this one and related stuff like conversions between representations.

**Comment by edgarriba** [comment]
@versatran01 please go ahead - more than happy that you can leverage this module.
Are you planning to base on this ? https://github.com/versatran01/liegroups

Few references:
- @krrish94 have already proposed some codes here: https://github.com/kornia/kornia/tree/sfm/kornia/sfm
- be aware also about this: https://github.com/kornia/kornia/issues/723

I think fist we need a clear API definition since I notice that the different implementation for liegroups are more OOP while in kornia we follow more a functional paradigm.

@versatran01 are you in our Slack - we could sync there with @krrish94 since I believe he's planning if not yet to use in [`gradslam`](https://github.com/gradslam/gradslam)

**Comment by versatran01** [comment]
I would start by proposing the following concepts and APIs.
This is by no means complete, but serves a good starting point.

A new module(file) named `rotation.py`, which can live wherever you see fit. For example `kornia.geometry.rotation`.
Inside we have the following concepts:
order: full name, short name, type, shape, representation
* `rotation_matrix`, `rmat`, `torch.tensor`, `(*,3,3)`, `[[r11, r12, r13], [r21, r22, r23], [r31, r32, r33]]`
The only issue is it can become non-orthonormal, and is hard to renormalize.
So we only use it as an intermediate representation (to transform points).
* `unit_quaternion`, `quat`, `torch.tensor`, `(*,4)`, `[qw, qx, qy, qz]`
Have the same issue with rotation matrix, where it needs to be re-normalized.
But at least its an easy operation.
The question here is for any function that takes `quat`, should we provide an option to re-normalize it?
More of this see below.
* `rotation_vector`, `rvec`, `torch.tensor`, `(*,3)`, `[rx, ry, rz]`
* `angle_axis`, `anax`, `torch.tensor`, `(*,4)`, `[an, ax, ay, az]`
Rotation vector is very similar to angle axis.
Since angle axis is indistinguishable from quaterion by shape (they both have 4 elements), 
I think we should only keep rotation vector for public api?
We could still use angle axis internally for convenience.
There are details to sort out when the magnitude of rotation vector is very small.
* `euler_angles`, `erpy`, `torch.tensor`, `(*,4)`, `[roll, pitch, yaw]`
We should not use this representation at all, just causes more confusion. 
List here for completeness.

We then provide the following conversions (could use full name but I use short name here).
The conversion function follows the form `a_from_b` such that you can write
`a = a_from_b(b_from_c(c))` instead of `a = b2a(c2b(c))`, which is just not intuitive.
For example `quat_from_rmat` and `rmat_from_quat`, and all the othe rcombinations.

**Comment by edgarriba** [comment]
@versatran01 this is a wonderful proposal - I like it.

I don't see any objection, everything you expose makes total sense to me. At some extend we followed a similar convention at least for the warping transformations since the initial contributor team had robotics background.

To not mess it up with the existing functionality, your proposal indeed needs a good planning so that we can keep backward compatibility. Probably the most difficult part will be the naming and code organization :)

Could you join our [slack](https://join.slack.com/t/kornia/shared_invite/zt-csobk21g-CnydWe5fmvkcktIeRFGCEQ) for more discussion or send me a pm to my email ? edgar at kornia.org

In the meantime, for developing I suggest you to read our [CONTRIBUTING](https://github.com/kornia/kornia/blob/master/CONTRIBUTING.rst) guidelines.

**Comment by edgarriba** [comment]
@versatran01 Any progress on this ?

**Comment by cjpurackal** [comment]
closed via #1887
